src/gui/qgsfilecontentsourcelineedit.h

```diff
@@ -0,0 +1,223 @@
+ * urls.
+ *
+ * \since QGIS 3.6
+ */
+#ifndef SIP_RUN
+class GUI_EXPORT QgsAbstractFileContentSourceLineEdit : public QWidget
+{
+#else
+class GUI_EXPORT QgsAbstractFileContentSourceLineEdit : public QWidget SIP_ABSTRACT
+{
+#endif
+    Q_OBJECT
+    Q_PROPERTY( QString source READ source WRITE setSource NOTIFY sourceChanged )
+
+  public:
+
+    /**
+     * Constructor for QgsAbstractFileContentSourceLineEdit, with the specified \a parent widget.
+     */
+    QgsAbstractFileContentSourceLineEdit( QWidget *parent SIP_TRANSFERTHIS = nullptr );
+
+    /**
+     * Returns the current file source.
+     * \see setSource()
+     * \see sourceChanged()
+     */
+    QString source() const;
+
+    /**
+     * Sets a specific settings \a key to use when storing the last
+     * used path for the file source.
+     */
+    void setLastPathSettingsKey( const QString &key );
+
+  public slots:
+
+    /**
+     * Sets a new \a source to show in the widget.
+     * \see source()
+     * \see sourceChanged()
+     */
+    void setSource( const QString &source );
+
+  signals:
+
+    /**
+     * Emitted whenever the file source is changed in the widget.
+     */
+    void sourceChanged( const QString &source );
+
+  private:
+
+#ifndef SIP_RUN
+///@cond PRIVATE
+
+    /**
+     * Returns the widget's file filter string.
+     */
+    virtual QString fileFilter() const = 0;
+
+    /**
+     * Returns the translated title to use for the select file dialog.
+     */
+    virtual QString selectFileTitle() const = 0;
+
+    /**
+     * Returns the translated title to use for the file from URL dialog.
+     */
+    virtual QString fileFromUrlTitle() const = 0;
+
+    /**
+     * Returns the translated descriptive text to use for the file from URL dialog.
+     */
+    virtual QString fileFromUrlText() const = 0;
+
+    /**
+     * Returns the translated title to use for the embed file dialog.
+     */
+    virtual QString embedFileTitle() const = 0;
+
+    /**
+     * Returns the translated title to use for the extract file dialog.
+     */
+    virtual QString extractFileTitle() const = 0;
+
+    /**
+     * Returns the default settings key to use for the widget's settings.
+     */
+    virtual QString defaultSettingsKey() const = 0;
+
+///@endcond
+#endif
+
+  private slots:
+    void selectFile();
+    void selectUrl();
+    void embedFile();
+    void extractFile();
+    void mFileLineEdit_textEdited( const QString &text );
+    void mFileLineEdit_editingFinished();
+
+  private:
+
+    QLineEdit *mFileLineEdit = nullptr;
+    QToolButton *mFileToolButton = nullptr;
+    QString mLastPathKey;
+    QString mBase64;
+
+    QString defaultPath() const;
+    QString settingsKey() const;
+
+};
+
+/**
+ * \ingroup gui
+ * \class QgsSvgSourceLineEdit
+ * A line edit widget with toolbutton for setting an SVG image path.
+ *
+ * Designed for use with QgsSvgCache.
+ *
+ * \see QgsImageSourceLineEdit
+ *
+ * \since QGIS 3.4
```

Comment:
3.4?

Reply:
Yes, that's in 3.4. it's just been moved/refactored here.